### PR TITLE
Remove unnecessary copies by using const ref or move

### PR DESCRIPTION
**Pull request by klecki**
Based on clang-tidy performance-* checks.
Mostly fixes for some string passing and TypeInfo.
Format argument.h to current style.

Signed-off-by: Krzysztof Lecki <klecki@nvidia.com>

### CODE REVIEW

File: dali/pipeline/operators/argument.h [hunk 1]
```diff
@@ -73,20 +65,20 @@ inline std::unique_ptr<Value> Value::construct(const T& val) {
...
 
-#define INSTANTIATE_VALUE_AS_INT64_PRESERVE_TYPE(T)                  \
-  template<>                                                         \
-  inline std::unique_ptr<Value> Value::construct(const T& val) {     \
-    auto ret = std::unique_ptr<Value>(new ValueInst<Index>(val));    \
-    /* preserve type information */                                  \
-    ret->SetTypeID(TypeTable::GetTypeID<T>());                       \
-    return ret;                                                      \
+#define INSTANTIATE_VALUE_AS_INT64_PRESERVE_TYPE(T)               \
+  template <>                                                     \
```
Comment (szalpal) [inline dali/pipeline/operators/argument.h]:
```suggestion
  template<>                                                     \
```
:p

**Comment by klecki** [comment]
Build: 674607

File: dali/pipeline/data/meta.h [hunk 2]
```diff
@@ -40,10 +40,14 @@ class DALIMeta {
     return source_info_;
   }
 
-  inline void SetSourceInfo(std::string source_info) {
+  inline void SetSourceInfo(const std::string &source_info) {
     source_info_ = source_info;
   }
 
+  inline void SetSourceInfo(std::string &&source_info) {
+    source_info_ = std::move(source_info);
```
Comment (Kh4L) [inline dali/pipeline/data/meta.h]: Why not having std::forward here and letting it determine the value category of the arg?

File: dali/pipeline/operators/argument.h [hunk 3]
```diff
@@ -56,13 +52,9 @@ class ValueInst : public Value {
     this->type_ = TypeTable::GetTypeID<T>();
   }
 
-  std::string ToString() const override {
-    return to_string(val_);
-  }
+  std::string ToString() const override { return to_string(val_); }
```
Comment (jantonguirao) [inline dali/pipeline/operators/argument.h]: I thought you didn't like one liners like that :)

File: dali/pipeline/data/meta.h (hunk 2 above)
Comment (klecki) [reply dali/pipeline/data/meta.h]: Exactly.

File: dali/pipeline/operators/argument.h (hunk 3 above)
Comment (klecki) [reply dali/pipeline/operators/argument.h]: Yup, auto-format did. Should probably unwrap it :P

File: dali/pipeline/data/meta.h (hunk 2 above)
Comment (Kh4L) [reply dali/pipeline/data/meta.h]: Oh right

**Comment by klecki** [comment]
!build

**Comment by klecki** [comment]
!build

File: dali/pipeline/operators/argument.h (hunk 3 above)
Comment (jantonguirao) [reply dali/pipeline/operators/argument.h]: ping

File: dali/pipeline/operators/decoder/cache/image_cache_largest.cc [hunk 4]
```diff
@@ -59,7 +59,7 @@ void ImageCacheLargest::Add(const ImageKey& image_key,
             && biggest_images_total_ + data_size > cache_size_
             && biggest_images_.top().first < data_size) {
           biggest_images_total_ -= biggest_images_.top().first;
-          to_be_discarded.push(std::move(biggest_images_.top()));
+          to_be_discarded.push(biggest_images_.top());
```
Comment (jantonguirao) [inline dali/pipeline/operators/decoder/cache/image_cache_largest.cc]: why is that?

Comment (klecki) [reply dali/pipeline/operators/decoder/cache/image_cache_largest.cc]: It's const&, you cannot move it.

File: dali/pipeline/operators/argument.h (hunk 3 above)
Comment (klecki) [reply dali/pipeline/operators/argument.h]: Done

**Comment by klecki** [comment]
!build

**Pull request merged**